engines/sherlock/screen.cpp

```diff
@@ -0,0 +1,534 @@
+	Events &events = *_vm->_events;
+
+	byte table[SHERLOCK_SCREEN_WIDTH];
+	Common::fill(&table[0], &table[SHERLOCK_SCREEN_WIDTH], 0);
+
+	for (int yp = 0; yp < SHERLOCK_SCREEN_HEIGHT; ++yp) {
+		for (int xp = 0; xp < SHERLOCK_SCREEN_WIDTH; ++xp) {
+			int temp = (table[xp] >= 197) ? SHERLOCK_SCREEN_HEIGHT - table[xp] :
+				_vm->getRandomNumber(3) + 1;
+
+			if (temp) {
+				blitFrom(_backBuffer1, Common::Point(xp, table[xp]),
+					Common::Rect(xp, table[xp], xp + 1, table[xp] + temp));
+				table[xp] += temp;
+			}
+		}
+
+		events.delay(10);
+	}
+}
+
+/**
+ * Copies a section of the second back buffer into the main back buffer
+ */
+void Screen::restoreBackground(const Common::Rect &r) {
+	if (r.width() > 0 && r.height() > 0) {
+		Common::Rect tempRect = r;
+		tempRect.clip(Common::Rect(0, 0, SHERLOCK_SCREEN_WIDTH, SHERLOCK_SCENE_HEIGHT));
+
+		if (tempRect.isValidRect())
+			_backBuffer1.blitFrom(_backBuffer2, Common::Point(tempRect.left, tempRect.top), tempRect);
+	}
+}
+
+/**
+ * Copies a given area to the screen
+ */
+void Screen::slamArea(int16 xp, int16 yp, int16 width, int16 height) {
+	slamRect(Common::Rect(xp, yp, xp + width, yp + height));
+}
+
+/**
+ * Copies a given area to the screen
+ */
+void Screen::slamRect(const Common::Rect &r) {
+	if (r.width() && r.height() > 0) {
+		Common::Rect tempRect = r;
+		tempRect.clip(Common::Rect(0, 0, SHERLOCK_SCREEN_WIDTH, SHERLOCK_SCREEN_HEIGHT));
+
+		if (tempRect.isValidRect())
+			blitFrom(*_backBuffer, Common::Point(tempRect.left, tempRect.top), tempRect);
+	}
+}
+
+/**
+ * Copy an image from the back buffer to the screen, taking care of both the
+ * new area covered by the shape as well as the old area, which must be restored
+ */
+void Screen::flushImage(ImageFrame *frame, const Common::Point &pt,
+		int16 *xp, int16 *yp, int16 *width, int16 *height) {
+	Common::Point imgPos = pt + frame->_offset;
+	Common::Rect newBounds(imgPos.x, imgPos.y, imgPos.x + frame->_frame.w, imgPos.y + frame->_frame.h);
+	Common::Rect oldBounds(*xp, *yp, *xp + *width, *yp + *height);
+
+	// See if the areas of the old and new overlap, and if so combine the areas
+	if (newBounds.intersects(oldBounds)) {
+		Common::Rect mergedBounds = newBounds;
+		mergedBounds.extend(oldBounds);
+		mergedBounds.right += 1;
+		mergedBounds.bottom += 1;
+
+		slamRect(mergedBounds);
+	} else {
+		// The two areas are independent, so copy them both
+		slamRect(newBounds);
+		slamRect(oldBounds);
+	}
+
+	*xp = newBounds.left;
+	*yp = newBounds.top;
+	*width = newBounds.width();
+	*height = newBounds.height();
+}
+
+/**
+ * Prints the text passed onto the back buffer at the given position and color.
+ * The string is then blitted to the screen
+ */
+void Screen::print(const Common::Point &pt, byte color, const char *formatStr, ...) {
+	// Create the string to display
+	char buffer[100];
+	va_list args;
+
+	va_start(args, formatStr);
+	vsprintf(buffer, formatStr, args);
+	va_end(args);
+	Common::String str(buffer);
```

Comment:
I strongly suggest to use `Common::String::vformat` instead of using the potentially unsafe vsprintf. Especially, since you end up putting this into a `Common::String` anyway.